### ISSUE TITLE
autograd_not_implemented_fallback registers a new warn hook on every call → unbounded hook list growth, backward slows down linearly

### ISSUE THREAD

**Issue by AkiRusProd** [pytorch/pytorch#191191]
### 🐛 Describe the bug

When an op has no autograd kernel and is called with inputs that require grad while grad mode is enabled, the autograd fallback (`AutogradFallbackMode::Warn`, the default) registers a **fresh** warn hook on every output tensor that requires grad — **on every single call**:

https://github.com/pytorch/pytorch/blob/v2.10.0/torch/csrc/autograd/autograd_not_implemented_fallback.cpp#L193-L196

```cpp
if (t.requires_grad()) {
  t.register_hook([op_name](const at::Tensor& grad) {
    warnAutogradNotImplemented(op_name);
  });
```

The returned handle is discarded, and there is no dedup: hook *N+1* is added even though the tensor already carries *N* identical warn hooks. For an in-place op on an `nn.Parameter` (e.g. `dist.broadcast(param)` → `c10d::broadcast_`) the output *is* the parameter, so the hooks pile up on a long-lived tensor that never goes away.

Those hooks live in the tensor's `cpp_hooks_list_`, which is exactly what `AccumulateGrad` iterates as its tensor pre-hooks:

https://github.com/pytorch/pytorch/blob/v2.10.0/torch/csrc/autograd/functions/accumulate_grad.h#L45-L51

So **every backward pass walks the entire accumulated list** and re-emits (well, re-checks) the same warning once per hook. Cost of backward grows linearly with the number of calls made to the fallback op since process start.

Two properties make this very hard to diagnose from user code:

1. **The warning hides its own scale.** `warnAutogradNotImplemented` goes through Python's warn-once dedup, so the user sees the deprecation message *once* and reasonably concludes it is informational — while the hooks behind it keep executing millions of times. (This is precisely what happened in #122521, closed by the reporter after confirming loss curves were unaffected.)
2. **The hooks are invisible to every user-side tool.** `cpp_hooks_list_` is a C++ `std::vector<std::function>`: `param._backward_hooks` stays empty, no Python objects are created (flat `gc.get_objects()` count, flat `tracemalloc`), and no Python frames appear in a sampling profiler. The only symptom is that the pre-hook zone of every `AccumulateGrad` node in a profiler trace keeps growing.

### Related, arguably a second bug: `remove_hook` never shrinks the list

Even code that dutifully saves and removes the handle would still pay a growing traversal cost, because removal only nulls the slot:

https://github.com/pytorch/pytorch/blob/v2.10.0/torch/csrc/autograd/variable.cpp#L739-L747

```cpp
void VariableHooks::remove_hook(const at::TensorBase& self, unsigned pos) const {
  auto& list = torch::autograd::impl::materialize_autograd_meta(self)->cpp_hooks_list_;
  TORCH_CHECK(list && pos < list->size(), "Invalid index, no hook at position ", pos);
  // Hook will be ignored
  (*list)[pos] = nullptr;
}
```

(Slots are never reclaimed, since positions are indices handed out to `RemovableHandle`s. Fine for a handful of hooks, unbounded for a register-per-call pattern.)

## Repro

CPU-only, `world_size=1`, gloo. `c10d::broadcast_` is used because it is the realistic trigger; any op that hits the fallback in a loop behaves the same.

```python
import os
import time
import warnings

import torch
import torch.distributed as dist

os.environ.setdefault("MASTER_ADDR", "127.0.0.1")
os.environ.setdefault("MASTER_PORT", "29517")
dist.init_process_group("gloo", rank=0, world_size=1)

N_PARAMS, DIM = 16, 64


def backward_us(params, iters=200):
    """Min backward wall time; dominated by the params' AccumulateGrad nodes."""
    best = float("inf")
    for _ in range(iters):
        loss = sum((p * 1.0).sum() for p in params)
        t0 = time.perf_counter()
        loss.backward()
        best = min(best, time.perf_counter() - t0)
        for p in params:
            p.grad = None
    return best * 1e6


def broadcast_n(params, n, no_grad):
    with warnings.catch_warnings():
        warnings.simplefilter("ignore")  # the fallback warns only once per location
        with torch.no_grad() if no_grad else torch.enable_grad():
            for _ in range(n):
                for p in params:
                    dist.broadcast(p, src=0)


print(f"torch {torch.__version__}")
for no_grad in (False, True):
    params = [torch.nn.Parameter(torch.randn(DIM)) for _ in range(N_PARAMS)]
    print(f"\nbroadcast under {'no_grad' if no_grad else 'grad mode (default)'}")
    print(f"{'broadcasts/param':>17} {'backward (min), us':>19} {'_backward_hooks':>16}")
    done = 0
    for total in (0, 100, 300, 1000):
        broadcast_n(params, total - done, no_grad)
        done = total
        print(f"{total:>17} {backward_us(params):>19.1f} "
              f"{len(params[0]._backward_hooks or ()):>16}")

dist.destroy_process_group()
```

Result on `2.12.0+cpu` (16 params of 64 floats — a toy model; a real model pays this per parameter):

| broadcasts/param | backward (min), µs — grad mode | backward (min), µs — under `no_grad` |
|---:|---:|---:|
| 0 | 469 | 444 |
| 100 | 3 224 | 442 |
| 300 | 10 283 | 445 |
| 1 000 | **36 334 (×77)** | **407 (flat)** |

Backward time is linear in the number of preceding fallback calls. The control column shows the growth disappears entirely when the broadcast runs under `no_grad`, i.e. when the fallback is not entered. `param._backward_hooks` is `0` in every row of both columns — nothing about this is observable from Python.

## Why this matters in practice

This is not a synthetic pattern. Any training loop that re-exports weights every step — e.g. Megatron-style weight synchronization, which broadcasts **raw** `named_parameters()` with `requires_grad=True` and grad mode on — adds +1 warn hook per parameter per step, forever.

We hit it as a mystery "training gets linearly slower, resets on restart" regression in a large distributed training job, and it took a long time to attribute, because every Python-level tool says the process is healthy. Concretely, after localizing it with a profiler:

- only the `AccumulateGrad` pre-hook zone grew — 18.6 → 32.3 → 54.0 µs per node between process-age 15, 60 and 120 optimizer steps, i.e. **+0.33 µs per node per step**, uniformly across all parameters, with node bodies and post-hooks constant;
- growth was a pure function of *process age*, independent of data or training stage (confirmed across three process incarnations at matched global steps);
- wrapping the broadcast in `torch.no_grad()` flattened the `AccumulateGrad` cost (11.4 µs at all three profiling points) and removed the step-time growth (~19× reduction in slope).

The user-side fix is trivial once you know (`no_grad`, or broadcast `param.data`), but the failure mode is essentially undiagnosable from Python, and the one visible clue — the deprecation warning — actively suggests it is harmless.

## Suggested fixes

Any of these would remove the unbounded growth:

1. **Register the warn hook at most once per tensor** (per op name, or even just once) — e.g. a flag in `AutogradMeta`, or check whether the same warn hook is already present. The warning semantics are unchanged: the hook is a warn-once notification, so *N* copies carry no more information than one.
2. **Don't register a hook on leaves at all**; the intent (warn if someone backprops through this) is already served for non-leaf outputs by the `WarnNotImplemented` grad_fn, and for a leaf the warning could be emitted eagerly at call time. This is the only one of the three that also fixes #134945.
3. **Make the fallback's warning non-hook-based** (warn at call time under warn-once dedup). Loses the "only warn if actually backpropped through" precision, but is O(1).

Independently: `remove_hook` should compact or erase the entry (e.g. shrink trailing nulls, or move `cpp_hooks_list_` to a map keyed by handle id) so that register/remove cycles don't grow the traversal cost either.

Also worth considering: registering an autograd kernel for the c10d in-place collectives (or at least having `dist.broadcast` do the broadcast under `no_grad` internally, since the op is not differentiable anyway) — that would fix the by-far most common way users hit this. Note that `dist.broadcast` on an `nn.Parameter` already has a second, separately-reported consequence via the same code path: #134945.

## Bonus: this also explains #134945 (silently dropped user hooks)

The per-call `register_hook` has a second, worse consequence, and it accounts for the open report in #134945 ("Broadcasting Parameter destroys hooks"). On a tensor whose `cpp_hooks_list_` is still null, `_register_hook` routes through `create_cpp_hook`, which **clears the tensor's existing hooks** and swaps in a fresh list:

https://github.com/pytorch/pytorch/blob/v2.10.0/torch/csrc/autograd/variable.cpp#L253-L268

```cpp
void create_cpp_hook(const at::TensorBase& self, bool is_retains_grad_hook) {
  const auto& fn = self.grad_fn();
  std::shared_ptr<hooks_list>& list = materialize_autograd_meta(self)->cpp_hooks_list_;
  list = std::make_shared<hooks_list>();
  ...
  clear_hooks(self);                                    // <-- drops what was there
  add_hook(self, std::make_unique<CppFunctionTensorPreHook>(list, 0));
```

So the *first* time the fallback fires on a parameter, it wipes the hook wiring the user had set up. Observed on `2.12.0+cpu` (`w = nn.Parameter(torch.ones(1))`, hook registered via `w.register_hook`, `w.backward()`):

| case | user hook fires |
|---|---|
| no broadcast (control) | ✅ |
| hook, then `dist.broadcast(w)` | ❌ |
| hook, then broadcast under `no_grad` | ✅ |
| broadcast, then hook | ✅ |
| broadcast, then hook, then 1 more broadcast | ✅ |
| broadcast, then hook, then 5 more broadcasts | ✅ |

Only the **first** broadcast on a given tensor destroys hooks; later ones are harmless. That is exactly the `if (!list)` guard in `_register_hook` — i.e. the damage is done by the `nullptr → list` transition inside `create_cpp_hook`, and the fallback is what triggers it without any user code asking for a C++ hook.

The failure is completely silent: after the broadcast, `w._backward_hooks` still contains the user's hook and `w.grad_fn` is still `None` (it is still a leaf) — only the C++-side wiring that connected that dict to the graph is gone.

Note this means suggestion 1 below (dedup) fixes the slowdown but **not** #134945: the very first registration is the destructive one.

## Related

- #122521 — `c10d::broadcast_: an autograd kernel was not registered...` with Megatron/fairscale TP>1. Exactly this code path; closed by the reporter as informational after checking loss curves. This issue is the argument that it isn't.
- #134945 — `dist.broadcast` on an `nn.Parameter` destroys gradient hooks (open, `oncall: distributed`). Same fallback; mechanism above.
- #146565 — `register_full_backward_hook` with c10d under `torch.compile` (adjacent).

## Versions (issue)

Numbers above measured on `2.12.0+cpu` (Linux, Python 3.11); also reproduced earlier on a 2.13 CPU build. Mechanism audited line-by-line on `v2.10.0`, and the same code is present on `main` at the time of writing. Nothing in the repro is CUDA- or backend-specific — it is the fallback, not the collective.

### Versions

PyTorch version: 2.12.0+cpu
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.5 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04.2) 11.4.0
Clang version: Could not collect
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.11.15 | packaged by conda-forge | (main, Mar  5 2026, 16:45:40) [GCC 14.3.0] (64-bit runtime)
Python platform: Linux-6.8.0-x86_64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Byte Order:                              Little Endian
Vendor ID:                               GenuineIntel
Model name:                              Intel(R) Xeon(R) Gold 6238 CPU @ 2.10GHz
CPU family:                              6
Model:                                   85
Stepping:                                7
Thread(s) per core:                      2
Core(s) per socket:                      22
Socket(s):                               4
NUMA node(s):                            4
Virtualization:                          VT-x
L1d cache:                               2.8 MiB
L1i cache:                               2.8 MiB
L2 cache:                                88 MiB
L3 cache:                                121 MiB

Versions of relevant libraries:
[pip3] numpy==2.4.4
[pip3] torch==2.12.0+cpu
[pip3] torchaudio==2.11.0+cpu
[pip3] torchvision==0.27.0+cpu
[conda] numpy                        2.4.4            pypi_0                pypi
[conda] torch                        2.12.0+cpu       pypi_0                pypi
[conda] torchaudio                   2.11.0+cpu       pypi_0                pypi
[conda] torchvision                  0.27.0+cpu       pypi_0                pypi

(This is `collect_env` output with the host-specific parts of the CPU section trimmed —
per-NUMA-node CPU id lists, the online/offline CPU lists, the ISA flag list and the
vulnerability-mitigation table — since this is a shared multi-socket host and the bug is
pure autograd bookkeeping in a single CPU-only process. Happy to provide any specific
field on request.)

cc @ezyang @albanD @gqchen @nikitaved @soulitzer @Varal7 @bobrenjc93

**Comment by eslam-ahmed43** [comment]
Hi, I would like to work on this issue if it is still available. I will look into it and open a PR when ready.
